Added support for template dir in `conan new`

TheQwertiest opened the pull request:
Changelog: Feature: Added support for template dir in `conan new`.
Docs: https://github.com/conan-io/docs/pull/1752

Implements/Fixes #6346 

Added a new `--template-dir` argument to `conan new`.

Usage:
```
conan new [-md TEMPLATE_DIR, --template-dir TEMPLATE_DIR] name
```

```
--template-dir    Use the given template directory from the local cache to generate a conan project
```

Template directory functions similarly to `--template TEMPLATE`, but takes directory as input instead. All files in this directory, their paths and names are processed via jinja.

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

TheQwertiest committed:
Added support for template dir in `conan new`

TheQwertiest commented:
Things to improve in the future:
- `--overwrite-policy {overwrite, ask, skip, abort}`: this will allow to seamlessly add conan support to existing projects.
- some sort of config file to list files to copy without rendering.

TheQwertiest commented:
Implementation limitation: currently all files are rendered and stored in memory (i.e. not written immediately after rendering), which is kinda dirty, and might be problematic if there are really big files.

TheQwertiest commented:
Hm... Now that I think about it... This functionality could be merged in the existing `--template` arg: if `TEMPLATE` is file - copy it, otherwise use the `--template-dir` behaviour. Should I do that?

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -289,6 +316,17 @@ def cmd_new(ref, header=False, pure_c=False, test=False, exports_sources=False,
         files = {"conanfile.py": replaced}
+    elif template_dir:
+        if not os.path.isabs(template_dir):
+            template_dir = os.path.join(cache.cache_folder, "templates", template_dir)
```
I'd really want to think about this. We already have `templates/outputs` for the table and graph templates, I would vote to add something like `commands/new` to this path and maybe the path from the command line (if relative) can start to look for the files inside this directory, wdyt? 

This is only a suggestion, I want to see other alternatives

TheQwertiest replied:
> I would vote to add something like commands/new to this path and maybe the path from the command line (if relative) can start to look for the files inside this directory, wdyt?

Sorry, I've failed to parse this sentence, could you rephrase this plz? What do you mean by "add something like commands/new to this path"?

jgsogo replied:
Sorry, I thought mine was a PR already released, but it isn't. We are introducing a way for the user to override the templates for the search table output and the graph info. Here you can see the PR to the docs: https://github.com/conan-io/docs/pull/1739.

Those templates are expected to be inside the directory `<cache>/templates/output/`. I expect more templates to arrive (like this PR you are suggesting here), so we need to organize that `templates` folder. My suggestion is to store templates for the `conan new` command inside a `<cache>/templates/commands/new/` directory.

The user, then, will use this new feature like:

* with an absolute path, everything clear:
   ```
   conan new --template /abs/path/to/folder
   ```
* with a relative one:
   ```
   conan new --template path/to/folder
   ```

   Now Conan can look first inside the working directory `$(pwd)/path/to/folder` and then inside the cache folder `<cache>/templates/commands/new/path/to/folder`.

   If using just a single name, it would be easier to understand, for example:

   ```
   conan new --template header_only
   ```
   
   Will get templates from `<cache>/templates/commands/new/header_only/`


I'm open to other alternatives, this is just a suggestion.

TheQwertiest replied:
Seems reasonable. Though I'm not sure that `commands` subdir is needed (i.e. `cmd_new` would be fine), since AFAIK there are no other conan commands that can (or could) use templates.

Btw, what about plain `.py` templates? Should they go into that new folder as well? Or maybe they should be deprecated altogether in favor of directory based templates?

jgsogo replied:
We need to keep supporting files as they were working before, but if something is better and breaking we can consider it as we are already designing the new Conan 2.0.

TheQwertiest replied:
fixed: 
- used `cmd_new` dir name
- added support for new subfolder in the old `.py` template case

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -236,13 +236,38 @@ def test(self):
+    if not os.path.isfile(os.path.join(template_dir, "conanfile.py")):
+        raise ConanException("Template is missing 'conanfile.py': {}".format(template_dir))
+
+    render_args = {'name': name, 'version': version, 'package_name': package_name}
```
```suggestion
    context = {'name': name, 'version': version, 'package_name': package_name}
```

I'd add at least the Conan version to this context.

TheQwertiest replied:
Is it this one?
`from conans import __version__ as client_version`

jgsogo replied:
Yes, that one, probably wrapped in a `Version` object:

```python
from conans.client.tools.version import Version

context = {...., 'version': Version(client_version)}
```

maybe name it `conan`? or `client_version`? or `conan_version`?

TheQwertiest replied:
Fixed.
Used plain str, cause `Version(ver)` rendered as `Object...`

jgsogo replied:
Ok, we can change it for a `Version` object if needed, and add a `__str__` to the class. But that can wait 👍

jgsogo reviewed (commented):
This is one of the improvements I had in mind after the search table and the graph 😉 did you read my mind?

I'd use the same `--template` argument, as you say it is perfectly compatible.

Probably we should use an `Environment` to load these templates, that's something I have also in mind but it will require more work (and some refactor), for this PR adding the feature is enough.

TheQwertiest commented:
> I'd use the same --template argument, as you say it is perfectly compatible.

The only confusing(?) thing about it would be that:
- `--template=file.py`: copies `file.py` as `conanfile.py` to the current directory.
- `--template=dir`: copies *content* of `dir` to the current directory and not `dir` itself,

Would that be alright?

TheQwertiest committed:
Review fixes

- collapsed `template-dir` in `template`
- added `conan_version` rendering
- moved templates to `cmd_new`

TheQwertiest commented:
Collapsed `template-dir` in `template`

TheQwertiest committed:
Amended --template arg description

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -281,14 +311,29 @@ def cmd_new(ref, header=False, pure_c=False, test=False, exports_sources=False,
+                                        version=version, 
+                                        package_name=package_name)
+            files = {"conanfile.py": replaced}
+        else:
```
If it is a folder, I think it should be incompatible with other arguments, at least with `test` and `pure-c` too. Other arguments might be deprecated in the new CLI implementation.

TheQwertiest replied:
> If it is a folder, I think it should be incompatible with other arguments, at least with test and pure-c too

`pure-c` should be incompatible with `--template=FILE` as well (judging by the code). Will be amended.

Not sure about the `test` though. I mean a template might not provide a `test_package` and only provide recipe and project structure. Maybe just add a check that there is no `test_package` files already (and error out if that's not the case)?

TheQwertiest replied:
Same for `gitignore`: template *might* provide this file, but it might not as well.

jgsogo replied:
Ok, leave it as it is. The new CLI will need to decide about all these arguments. There are too many flags here.

TheQwertiest replied:
I've added a check for `pure-c` though =) (since it's ignored with `template` anyways)

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -281,14 +311,29 @@ def cmd_new(ref, header=False, pure_c=False, test=False, exports_sources=False,
+        is_file_template = os.path.basename(template).endswith('.py')
+        if is_file_template:
+            if not os.path.isabs(template):
+                old_path = os.path.join(cache.cache_folder, "templates", template)
```
Add a comment here, like

```
# FIXME: Conan 2.0. The old path should be removed
```

TheQwertiest replied:
Fixed

jgsogo commented on conans/client/command.py:
```diff
@@ -224,7 +224,9 @@ def new(self, *args):
                         circleci_osx_versions=args.ci_circleci_osx,
-                        template=args.template)
+                        template=args.template,
+                        template_dir=args.template_dir
```
There is no longer a `args.template_dir`, isn't it?  (Tests are failing here)

TheQwertiest replied:
Fixed

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -236,13 +237,42 @@ def test(self):
+
+def _get_files_from_template_dir(template_dir, name, version, package_name):
+    if not os.path.isfile(os.path.join(template_dir, "conanfile.py")):
+        raise ConanException("Template is missing 'conanfile.py': {}".format(template_dir))
```
IMO, we shouldn't add this check, the conanfile can take any other name, or maybe there is a `conanfile.txt`...

TheQwertiest replied:
I would really like to keep this type of check to prevent the following scenario: user makes a typo or template dir is malformed (e.g. `conan config install` did not go through), so the command `conan new --template=T` finishes successfully, but there are no files or worse a whole load of random files from the wrong location (the classic `cp -rf / path` mistake).

jgsogo replied:
If use enters a typo in `conan new --template=<typo>` it will fail because the folder/file doesn't exist. If there was a problem during the `conan config install` then, after this command runs successfully, you will realize about it in your working dir: missing files, a project that doesn't work... 

If you are worried about someone entering a relative path like `conan new --template=../../<typo>`, we can limit the search in the cache to folders inside the `<cache>/templates/cmd_new/` directory. It will work if it is found locally (`$(pwd)/../../<typo>`), but looking in the cache will raise.

TheQwertiest replied:
IIRC, `--template` accepts absolute paths as well. So, `conan new --template=/ path` will work (it will take huge time to complete though :D). This check will also prevent users from (ab)using templates as a storage for random files.

jgsogo replied:
Ok, that was my opinion, but we can wait for others to review and add their two cents 👍

TheQwertiest replied:
Well, at the end of the day, you are the dev.
I.e. if you think that this check is not worth having then it's gotta go =)

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -236,13 +237,42 @@ def test(self):
+        for f in fs:
+            rel_d = os.path.relpath(d, template_dir)
+            rel_f = os.path.join(rel_d, f)
+            files += [rel_f]
```
```suggestion
            files.append(rel_f)
```

jgsogo commented on conans/client/cmd/new.py:
```diff
@@ -236,13 +237,42 @@ def test(self):
+    out_files = dict()
+    for f in files:
+        f_path = os.path.join(template_dir, f)
+        if not os.path.isfile(f_path):
+            continue
```
```suggestion
```

This should never happen

jgsogo reviewed (commented):
Asking for some changes, but overall I really like it. Still I need to clone the branch and play with the feature, but 👌

TheQwertiest committed:
Update conans/client/cmd/new.py

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>

TheQwertiest committed:
Update conans/client/cmd/new.py

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>

TheQwertiest committed:
Review fixes

TheQwertiest commented:
@jgsogo , btw are conan docs generated dynamically (e.g. https://docs.conan.io/en/latest/reference/commands/creator/new.html)?
In other words, do I have to edit command output manually (by replacing it with the output of `conan new --help`) or do I leave it as is?

jgsogo commented:
The output is generated _automatically_ with a script. We run it manually, but it is scripted (https://github.com/conan-io/docs/blob/develop/commands_help_update.py).

memsharded commented on conans/client/conan_api.py:
```diff
@@ -264,7 +264,8 @@ def new(self, name, header=False, pure_c=False, test=False, exports_sources=Fals
                         circleci_clang_versions=circleci_clang_versions,
                         circleci_osx_versions=circleci_osx_versions,
-                        template=template, cache=self.app.cache)
+                        template=template, template_dir=template_dir, 
```
Shouldn't the template be a path, it can be a path to a file or a path to a folder? Actually it seems the arg is not used in the cmd new.py

TheQwertiest replied:
Fixed

memsharded commented on conans/client/cmd/new.py:
```diff
@@ -236,13 +237,41 @@ def test(self):
             gitlab_gcc_versions=None, gitlab_clang_versions=None,
             circleci_gcc_versions=None, circleci_clang_versions=None, circleci_osx_versions=None,
-            template=None, cache=None):
+            template=None, template_dir=None, cache=None):
```
Seems ``template_dir`` argument not used here?

TheQwertiest replied:
Fixed

TheQwertiest committed:
Review fixes

TheQwertiest commented:
> The output is generated automatically with a script. We run it manually

Uhm, so do *I* need to run it when making a doc PR? Or is it only invoked right before making a release?

jgsogo commented:
> > The output is generated automatically with a script. We run it manually
> 
> Uhm, so do _I_ need to run it when making a doc PR? Or is it only invoked right before making a release?

Better when making the PR, but TBH, we usually forget about it and we run the script just before making a release 😅

memsharded commented on conans/client/cmd/new.py:
```diff
@@ -281,14 +310,30 @@ def cmd_new(ref, header=False, pure_c=False, test=False, exports_sources=False,
+            if not os.path.isabs(template):
+                # FIXME: Conan 2.0. The old path should be removed
+                old_path = os.path.join(cache.cache_folder, "templates", template)
+                new_path = os.path.join(cache.cache_folder, "templates", "cmd_new", template)
```
Hi @TheQwertiest 

Please, lets use the path ``templates/command/new``. 

The reason is that in Conan 2.0 we are going to modularize the commands, making it an extensible framework to plug user commands. So it will be cleaner to have all of those things inside a ``templates/command/`` subfolder, then followed by the name of the command.

Many thanks!

TheQwertiest replied:
Sure

TheQwertiest replied:
Done.

PS: Sorry for nagging, but I would really appreciate if you answer a question about Python 2.7 in meson helper PR =(

memsharded reviewed (approved):
This is great, thanks very much.

I would like to merge it on Monday to be included in next 1.27. There is just a minor thing to improve with the template path, if you can address it by then, it would be great, if not we can do it ourselves so we don't delay this contribution any more.

Thanks again for your contribution!

TheQwertiest committed:
Review fixes

jgsogo reviewed (approved):
💯 !

We need docs for this new feature, they should document the fields that can be used in the template and include an example. Let us know if you want to write them ourselves or you will do it.

Thanks!

TheQwertiest commented:
Done: https://github.com/conan-io/docs/pull/1752
Note the note in that PR about cmd line output (pun intended :P)

memsharded commented:
Great, thanks for this contribution!
